### DISCUSSION TITLE
Tailwind and parcel: cannot get rid of "PostCSS plugin postcss-purgecss requires PostCSS 8."

### DISCUSSION THREAD

**Discussion by brpaz** [tailwindlabs/tailwindcss#3191] [Help]
Hello.
I am creating a sample site from scratch with Parcel bundler and Tailwind.

I have installed tailwind with PostCSS compact as specified in the docs but I always get the error when building:

```
🚨  /tmp/static-site/src/index.css:undefined:undefined: PostCSS plugin postcss-purgecss requires PostCSS 8.
Migration guide for end-users:
https://github.com/postcss/postcss/wiki/PostCSS-8-for-end-users
Migration guide for end-users:
https://github.com/postcss/postcss/wiki/PostCSS-8-for-end-users
    at Processor.normalize (/tmp/static-site/node_modules/postcss/lib/processor.js:167:15)
    at new Processor (/tmp/static-site/node_modules/postcss/lib/processor.js:56:25)
    at postcss (/tmp/static-site/node_modules/postcss/lib/postcss.js:55:10)
    at purgeUnusedUtilities (/tmp/static-site/node_modules/tailwindcss/lib/lib/purgeUnusedStyles.js:69:31)
    at /tmp/static-site/node_modules/tailwindcss/lib/processTailwindFeatures.js:75:555
    at LazyResult.run (/tmp/static-site/node_modules/postcss/lib/lazy-result.js:288:14)
    at LazyResult.asyncTick (/tmp/static-site/node_modules/postcss/lib/lazy-result.js:212:26)
    at LazyResult.asyncTick (/tmp/static-site/node_modules/postcss/lib/lazy-result.js:225:14)
    at /tmp/static-site/node_modules/postcss/lib/lazy-result.js:254:14
    at new Promise (<anonymous>)
error Command failed with exit code 1.
```

I had removed my node_modules folder and installed everything again but without success.
What I am missing?

Heere are the relevant contents of my configuration files.

**package.json**:

```json
  "dependencies": {
    "@tailwindcss/postcss7-compat": "^2.0.2",
    "autoprefixer": "^9",
    "postcss": "^7",
    "pug": "^2.0.4",
    "tailwindcss": "npm:@tailwindcss/postcss7-compat"
  },
```

**postcss.config.js**

```js
module.exports = {
  plugins: {
    tailwindcss: {},
    autoprefixer: {},
  },
};
```

**tailwind.config.js**

```js
module.exports = {
  theme: {
    extend: {},
  },
  variants: {},
  plugins: [],
  purge: ['./src/**/*.html', './src/**/*.pug'],
};
```

**Comment by trapcodeio** [comment]
Same here..

**Reply by ShahirZain** [reply]
sudo npm cache clean -f
sudo npm install -g n
sudo n stable
To upgrade to latest version (and not current stable) version, you can use
sudo n latest
sudo apt-get install --reinstall nodejs-legacy     # fix /usr/bin/node

### now do this

vue create demo

npm install tailwindcss

//Create a css folder inside assets directory. Then main.css file
cd src/assets && mkdir css && cd css && touch main.css

//Back to the root of the project
../../../

npx tailwindcss init

### Inside Main.css paste this

@tailwind base;
@tailwind components;
select {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  outline: none;
}

input {
  outline: none;
}

@tailwind utilities;


**Import the main.css inside the main.js file.**

**Then create the Post CSS configuration file.**
$ touch postcss.config.js

### Inside the postcss.config.js, copy and paste this:
const autoprefixer = require('autoprefixer');
const tailwindcss = require('tailwindcss');

module.exports = {
  plugins: [
    tailwindcss,
    autoprefixer,
  ],
};

### then final step do this
npm uninstall tailwindcss postcss autoprefixer
npm install tailwindcss@npm:@tailwindcss/postcss7-compat @tailwindcss/postcss7-compat postcss@^7 autoprefixer@^9

**Comment by trapcodeio** [comment]
Please any fix for this?

**Comment by raknjarasoa** [comment]
work well with `parcel@nightly` from [official support](https://github.com/postcss/postcss/wiki/PostCSS-8-for-end-users#parcel) or see https://tailwindcss.com/docs/installation#post-css-7-compatibility-build

**Comment by FDiskas** [comment]
nightly has a lot of other issues and this is not a solution. 
If you install ` "tailwindcss": "npm:@tailwindcss/postcss7-compat",`  you also get a dependency postcss v8
![image](https://user-images.githubusercontent.com/468006/103462748-13bf3880-4d30-11eb-8b26-c11910fd6fa7.png)

**Reply by Dimimo** [reply]
Upgrading NodeJS to version 14 fixed it for me. When done, I ran `npm outdated` and `npm upgrade -D` and no more issues.

**Reply by RichardJECooke** [reply]
Sadly didn't work for me

**Comment by lmuzquiz** [comment]
Not sure if this is related but had a similar error message on my console when running **nuxt generate** using nuxt/TailwindCSS v3.4.2 on Node v12

> ERROR in ./node_modules/@nuxtjs/tailwindcss/lib/files/tailwind.css (./node_modules/css-loader/dist/cjs.js??ref--3-oneOf-1-1!./node_modules/postcss-loader/src??ref--3-oneOf-1-2!./node_modules/@nuxtjs/tailwindcss/lib/files/tailwind.css)
Module build failed (from ./node_modules/postcss-loader/src/index.js):
Error: PostCSS plugin postcss-purgecss requires PostCSS 8.

**### What i did was to remove  **"@teamnovu/nuxt-breaky": "^1.2.2",** from my package.json, then **rm -rf the node_modules** folder (delete it) and run **yarn install again****

<img width="576" alt="Screen Shot 2021-01-05 at 19 59 03" src="https://user-images.githubusercontent.com/11865448/103720242-8af30780-4f90-11eb-9cdb-29d929c05d37.png">

This was part of the error:

<img width="1106" alt="Screen Shot 2021-01-05 at 19 52 50" src="https://user-images.githubusercontent.com/11865448/103720365-cd1c4900-4f90-11eb-8fbb-cd2cdf01d020.png">

"dependencies": {
    "core-js": "^3.6.5",
    "gsap": "./gsap-bonus.tgz",
    "nuxt": "^2.14.6"
  },
  "devDependencies": {
    "@nuxtjs/eslint-config": "^3.1.0",
    "@nuxtjs/eslint-module": "^2.0.0",
    "@nuxtjs/tailwindcss": "^3.4.2",
    "@teamnovu/nuxt-breaky": "^1.2.2",
    "autoprefixer": "^9",
    "babel-eslint": "^10.1.0",
    "eslint": "^7.10.0",
    "eslint-config-prettier": "^6.12.0",
    "eslint-plugin-nuxt": "^1.0.0",
    "eslint-plugin-prettier": "^3.1.4",
    "node-sass": "^5.0.0",
    "postcss": "^7",
    "prettier": "^2.1.2",
    "sass-loader": "^10.1.0",
    "tailwindcss": "npm:@tailwindcss/postcss7-compat"
  }

**Reply by FDiskas** [reply]
probably you updated some deep dependency - that was my issue also. Do the same but install with `yarn install --frozen-lockfile`

**Reply by lmuzquiz** [reply]
`yarn install --frozen-lockfile` did not worked for me

**Reply by FDiskas** [reply]
did your lock file also updated? search for related changes and revert them

**Comment by ganuonglachanh** [comment]
To prevent error "Error: PostCSS plugin postcss-purgecss requires PostCSS 8":
```
yarn remove tailwindcss postcss autoprefixer
yarn add tailwindcss@npm:@tailwindcss/postcss7-compat @tailwindcss/postcss7-compat postcss@^7 autoprefixer@^9
yarn -D @fullhuman/postcss-purgecss@3.0.0
```

**Reply by porterbayne** [reply]
Thank you -- I'm stuck with this on a deploy.  But, what is `yarn -D` ?  I can't get that to run.

**Reply by porterbayne** [reply]
(I did then just adjust the PurgeCSS version in package.json manually and that seems to have worked)

**Reply by Bluefitdev** [reply]
OMG this works... how is this not documented. I've been searching for hours for solutions! Thankyou!!

**Reply by FDiskas** [reply]
> Thank you -- I'm stuck with this on a deploy.  But, what is `yarn -D` ?  I can't get that to run.

yarn add -D *** installs module to dev dependecies

**Reply by aaditya-007** [reply]
Works fro me. Thanks

**Reply by robertwt7** [reply]
This should be documented.. somehow the last step is very important

**Reply by dannyrb** [reply]
This works, but it's a downgrade of `postcss-purgecss`?

**Reply by jlvallelonga** [reply]
in case anyone's using npm:
```
npm uninstall tailwindcss postcss autoprefixer
npm install tailwindcss@npm:@tailwindcss/postcss7-compat @tailwindcss/postcss7-compat postcss@^7 autoprefixer@^9
npm install -D @fullhuman/postcss-purgecss@3.0.0
```

**Reply by ashwinkk** [reply]
Thank you!

**Reply by haotan19** [reply]
Thanks!

But this solution only works for `yarn dev` or `parcel index.html`.
I try to use it on `yarn build` or `parcel build index.html`, and it gives me the same error:
> src/styles/globals.css:undefined:undefined: PostCSS plugin postcss-purgecss requires PostCSS 8. 

Any idea?
